src/client/batch.rs

```diff
@@ -0,0 +1,178 @@
+pub(crate) enum FlushResult<T> {
+    Ok(T, Option<JoinHandle<()>>),
+    Error(T, Error),
+}
+
+/// A [`BatchBuilder`] uses an [`Aggregator`] to construct maximally large batch
+/// of writes, and returning a [`ResultHandle`] for callers to demux the result.
+///
+/// Once flushed by a caller, the [`BatchBuilder`] spawns a task to
+/// asynchronously write the batch to Kafka, immediately returning a new
+/// instance of [`BatchBuilder`] to begin buffering up a new batch.
+#[derive(Debug)]
+pub(crate) struct BatchBuilder<A>
+where
+    A: Aggregator,
+{
+    aggregator: A,
+    results: BroadcastOnce<BatchWriteResult<A>>,
+}
+
+impl<A> BatchBuilder<A>
+where
+    A: Aggregator,
+{
+    pub(crate) fn new(aggregator: A) -> Self {
+        Self {
+            aggregator,
+            results: Default::default(),
+        }
+    }
+
+    /// Attempt to place this `data` into the underlying [`Aggregator`] impl.
+    ///
+    /// Returns a handle to obtain the write result if successful. If
+    /// [`TryPush::NoCapacity`] is returned, subsequent calls to
+    /// [`Self::try_push()`] will likely fail and this batch should be flushed
+    /// by calling [`Self::background_flush()`].
+    pub(super) fn try_push(
+        &mut self,
+        data: A::Input,
+    ) -> Result<TryPush<A::Input, ResultHandle<A>>, Error> {
+        match self
+            .aggregator
+            .try_push(data)
+            .map_err(|e| Error::Aggregator(e.into()))?
+        {
+            TryPush::NoCapacity(data) => Ok(TryPush::NoCapacity(data)),
+            TryPush::Aggregated(tag) => Ok(TryPush::Aggregated(ResultHandle::new(
+                self.results.receiver(),
+                tag,
+            ))),
+        }
+    }
+
+    /// Perform an asynchronous flush of this buffer.
+    ///
+    /// Returns a handle to the async flush task if a flush was necessary.
+    pub(super) fn background_flush(
+        mut self,
+        client: Arc<dyn ProducerClient>,
+        compression: Compression,
+    ) -> FlushResult<Self> {
+        let (batch, status_deagg) = match self.aggregator.flush() {
+            Ok(v) => v,
+            Err(e) => {
+                return FlushResult::Error(Self::new(self.aggregator), Error::Aggregator(e.into()))
+            }
+        };
+
+        if batch.is_empty() {
+            debug!(?client, "No data aggregated, skipping client request");
+            // A custom aggregator might have produced no records, but the
+            // the `produce()` callers are still waiting for their
+            // responses.
+            //
+            // Broadcast an empty result set to satisfy the aggregation
+            // contract.
+            self.results.broadcast(Ok(Arc::new(AggregatedStatus {
+                aggregated_status: vec![],
+                status_deagg,
+            })));
+            return FlushResult::Ok(Self::new(self.aggregator), None);
+        }
+
+        let handle = tokio::spawn({
```

Comment:
Why do we need yet another task here? Isn't there already one that drives the linger and the flushing? Or is this so we have also cancellation safety if there are manual flushes?